### PR TITLE
Reuse the authorization and template shared informers for GC

### PR DESCRIPTION
**Pull request by smarterclayton**
Allow them to be loaded via ForResource() as a stepping stone to
transforming their behavior.

For #8229

[test]

### CODE REVIEW

**Comment by smarterclayton** [comment]
@liggitt I am confused why the preferred group versions are changing

**Comment by liggitt** [comment]
I'm not sure either

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
It's because the generated informers require their API group to be registered.

**Comment by smarterclayton** [comment]
Bolt db segmentation fault [test]

**Comment by smarterclayton** [comment]
Fixed, review.

**Comment by smarterclayton** [comment]
@enj also for review

File: pkg/cmd/server/start/start_master.go [hunk 1]
```diff
@@ -580,6 +582,29 @@ func checkForOverrideConfig(ac configapi.AdmissionConfig) (*overrideapi.ClusterR
 	return overrideConfig, nil
 }
 
+type GenericResourceInformer interface {
+	ForResource(resource schema.GroupVersionResource) (kinformers.GenericInformer, error)
+}
+
+type genericInformers struct {
```
Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: This should be in a separate package.  I agree that combining informers is useful, but I suspect we'll need it more generally.

File: pkg/cmd/server/start/start_master.go [hunk 2]
```diff
@@ -580,6 +582,29 @@ func checkForOverrideConfig(ac configapi.AdmissionConfig) (*overrideapi.ClusterR
...
+	kinformers.SharedInformerFactory
```
Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: Why must this be anonymous?

File: pkg/cmd/server/start/start_master.go [hunk 3]
```diff
@@ -648,8 +673,11 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 	}
 
 	controllerContext := kctrlmgr.ControllerContext{
-		ClientBuilder:      saClientBuilder,
-		InformerFactory:    oc.Informers.KubernetesInformers(),
+		ClientBuilder: saClientBuilder,
+		InformerFactory: genericInformers{
+			SharedInformerFactory: oc.Informers.KubernetesInformers(),
+			generic:               []GenericResourceInformer{oc.Informers},
```
Comment (deads2k) [inline pkg/cmd/server/start/start_master.go]: Isn't this wiring in the hand coded informers?  I'm against including them in new places.

File: pkg/controller/shared/shared_informer.go [hunk 4]
```diff
@@ -81,13 +87,26 @@ type sharedInformerFactory struct {
 	customListerWatchers  ListerWatcherOverrides
 	defaultResync         time.Duration
 
+	templateInformers      templateinformers.SharedInformerFactory
```
Comment (deads2k) [inline pkg/controller/shared/shared_informer.go]: I don't like this. If we must have a new struct, we should be building out a new struct that doesn't carry the legacy of hand-coded informers with it.

File: pkg/cmd/server/start/start_master.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_master.go]: this is passing through all the kubernetes informer methods

File: pkg/cmd/server/start/start_master.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_master.go]: We need those to reuse existing informers

File: pkg/controller/shared/shared_informer.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/controller/shared/shared_informer.go]: What do you want the tsruct to look like?  The upstream shared struct?  This is the generated type

File: pkg/cmd/server/start/start_master.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/cmd/server/start/start_master.go]: This was just the adapter interface for GC, if you want to start a brand new controller factory i want to do that in a separate PR where the rest of them get converted.

**Comment by smarterclayton** [comment]
Refactored to be clearer, removed Register, and excluded a lot more resources that I honestly don't care about right now (mostly ones without reflectors).

**Comment by smarterclayton** [comment]
[merge] based on feedback from previous
